model/pod/dispatcher/dispatcher.go

```diff
@@ -74,3 +86,201 @@ func (pd *PodDispatcher) UpsertAtomically() (*adb.ChangeInfo, error) {
+		}
+
+		event.LogPodAssignedTask(p.ID, t.Id)
+		event.LogContainerTaskDispatched(t.Id, t.Execution, p.ID)
+
+		return t, nil
+	}
+
+	return nil, nil
+}
+
+// dispatchTaskAtomically performs the DB updates to assign a task to run on a
+// pod.
+func (pd *PodDispatcher) dispatchTaskAtomically(ctx context.Context, env evergreen.Environment, p *pod.Pod, t *task.Task) error {
+	session, err := env.Client().StartSession()
+	if err != nil {
+		return errors.Wrap(err, "starting transaction session")
+	}
+	defer session.EndSession(ctx)
+
+	if _, err := session.WithTransaction(ctx, pd.dispatchTask(env, p, t)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (pd *PodDispatcher) dispatchTask(env evergreen.Environment, p *pod.Pod, t *task.Task) func(mongo.SessionContext) (interface{}, error) {
+	return func(sessCtx mongo.SessionContext) (interface{}, error) {
+		if err := p.SetRunningTask(sessCtx, env, t.Id); err != nil {
+			return nil, errors.Wrapf(err, "setting pod's running task")
+		}
+
+		if err := t.MarkAsContainerDispatched(sessCtx, env, p.AgentVersion, time.Now()); err != nil {
+			return nil, errors.Wrapf(err, "marking task as dispatched")
+		}
+
+		if err := pd.dequeue(sessCtx, env); err != nil {
+			return nil, errors.Wrapf(err, "dequeueing task")
+		}
+
+		return nil, nil
+	}
+}
+
+// dequeue removes the head of the task queue from the dispatcher.
+func (pd *PodDispatcher) dequeue(ctx context.Context, env evergreen.Environment) (err error) {
+	oldTaskIDs := pd.TaskIDs
+	pd.TaskIDs = pd.TaskIDs[1:]
+	defer func() {
+		if err != nil {
+			pd.TaskIDs = oldTaskIDs
+		}
+	}()
+	res, err := env.DB().Collection(Collection).UpdateOne(ctx, pd.atomicUpsertQuery(), pd.atomicUpsertUpdate())
+	if err != nil {
+		return errors.Wrap(err, "upserting dispatcher")
+	}
+	if res.ModifiedCount == 0 {
+		return errors.New("dispatcher was not updated")
+	}
+
+	pd.ModificationCount++
+
+	return nil
+}
+
+// dequeueUndispatchableTask removes the task from the dispatcher and, if the
+// task status indicates a container has been allocated for it, marks it as
+// unallocated.
+func (pd *PodDispatcher) dequeueUndispatchableTask(ctx context.Context, env evergreen.Environment, t *task.Task) error {
+	if t.Status != evergreen.TaskContainerAllocated {
+		return errors.Wrap(pd.dequeue(ctx, env), "dequeueing task")
+	}
+
+	if err := t.MarkAsContainerUnallocated(ctx, env); err != nil {
+		return errors.Wrap(err, "marking task unallocated")
+	}
+
+	if err := pd.dequeue(ctx, env); err != nil {
+		return errors.Wrap(err, "dequeueing task")
+	}
+
+	return nil
+}
+
+// checkTaskIsDispatchable checks if a task is able to dispatch based on its
+// current state and its project ref's settings.
+func (pd *PodDispatcher) checkTaskIsDispatchable(ctx context.Context, env evergreen.Environment, t *task.Task) (shouldRun bool, err error) {
+	if !t.IsContainerDispatchable() {
+		grip.Notice(message.Fields{
+			"message":    "task in dispatch queue is not dispatchable",
```

Comment:
Nit: `is not container dispatchable`.

Reply:
I slightly modified it to say "container task". The container/host part is just to distinguish if it'll return the correct value for a container/host task.